src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            addFirst(value);
+        } else {
+            Node<T> node = new Node<>(tail, value, null);
+            tail.next = node;
+            tail = node;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            addFirst(value);
+        } else if (isIndexValid(index)) {
+            Node<T> currentNode = searchNodeByIndex(index);
+            Node<T> node = new Node<T>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = node;
+            currentNode.prev = node;
+            size++;
+        } else {
+            throw new IndexOutOfBoundsException("Index is invalid");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexValid(index)) {
+            return searchNodeByIndex(index).value;
+        } else {
+            throw new IndexOutOfBoundsException("Index is invalid");
```

Comment:
here, it seems logical, but I am not fully confident. Would you agree to remove this error in isIndexValid() method, so your exception wouldn't repeat?

Reply:
you are right!!! Thanks!